In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df=pd.read_csv('/content/crime_data (1).csv')

In [ ]:
y.unique()


array([ 761,  780,  593,  715, 1078,  567,  456,  686, 1206,  723,  261,
        326,  282,  960,  489,  496,  463, 1062,  805,  998,  126,  792,
        327,  744,  434,  178,  679,   82,  339,  138,  627,  930,  875,
       1074,  504,  635,  503,  418,  402, 1023,  208,  766,  762,  301,
        372,  114,  515,  264,  286, 2922])

In [ ]:
#eliminamos columnas que no necesitamos y dejamos las columnas asignadas al ultimo numero de mi matricula
#8 - Variable dependiente VR, variables independientes M, W, H, P y S
df.drop(columns=[df.columns[0], df.columns[2]], inplace=True)
df.head()

VR     M     W     H     P     S
0   761  41.8  75.2  86.6   9.1  14.3
1   780  67.4  73.5  66.9  17.4  11.5
2   593  44.7  82.9  66.3  20.0  10.7
3   715  84.7  88.6  78.7  15.4  12.1
4  1078  96.7  79.3  76.2  18.2  12.5

## 1. Evalúa con validación cruzada un modelo pereceptrón multicapa para las variables que se te asignaron para este ejercicio.

Utilizamos esta opcion

Modelo de regresión con red neuronal perceptrón multicapa:
porque la variable que queremos predecir  es continua y se busca predecir un valor numérico


In [ ]:
from sklearn import datasets
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [ ]:
y = df.iloc[:, 0]
x = df.iloc[:, 1:]
n_features = x.shape[1]
print(n_features)

targets = y.unique()
n_clases = len(targets)
n_clases

5


50

In [ ]:
# Fit regression model using the complete dataset and calculate MSE and MAE
regr = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
regr.fit(x, y)

y_pred = regr.predict(x)
print('MSE: ', mean_squared_error(y, y_pred))
print("MAE: ", mean_absolute_error(y, y_pred))

# Evaluate regression model using k-fold cross-validation
n_folds = 3
# Set a random seed for reproducibility
np.random.seed(42)
kf = KFold(n_splits=n_folds, shuffle = True,  random_state=42)

mse_cv = []
mae_cv = []
for train_index, test_index in kf.split(x):

    # Training phase
    x_train = x.iloc[train_index]
    y_train = y.iloc[train_index]

    regr_cv = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
    regr_cv.fit(x_train, y_train)

    # Test phase
    x_test = x.iloc[test_index]
    y_test = y.iloc[test_index]

    y_pred = regr_cv.predict(x_test)

    # Calculate MSE and MAE

    mse_i = mean_squared_error(y_test, y_pred)
    print('mse = ', mse_i)
    mse_cv.append(mse_i)

    mae_i = mean_absolute_error(y_test, y_pred)
    print('mae = ', mae_i)
    mae_cv.append(mae_i)



# Calculate and print the average MSE and MAE across all folds
avg_mse = np.mean(mse_cv)
print(f"Average MSE across all folds: {avg_mse}")
avg_msa = np.mean(mae_cv)
print(f"Average MAE across all folds: {avg_msa}")



MSE:  13180.29625575108
MAE:  88.9268495578019
mse =  71414.85603199447
mae =  173.83062761520995
mse =  38768.93931504232
mae =  160.910752822593
mse =  227054.3188347926
mae =  241.1539535433522
Average MSE across all folds: 112412.70472727646
Average MAE across all folds: 191.9651113270517


Un mayor valor promedio de MSE (112412.70) y MAE (191.97) en la validación cruzada k-fold en comparación con el MSE inicial (13180.30) y el MAE (88.93) sugieren que mi modelo podría no generalizar bien a datos no vistos. Esto significa que el rendimiento del modelo en los datos de entrenamiento parece ser mejor que su rendimiento en nuevos datos, lo que sugiere un posible problema de sobreajuste.

En otras palabras, los resultados de la validación cruzada indican que mi modelo no se desempeña tan bien en datos que no ha visto previamente como lo hace en los datos en los que fue entrenado. Esto podría deberse a que el modelo está demasiado ajustado a los datos de entrenamiento y no puede generalizar de manera efectiva a nuevos datos. Para mejorar el rendimiento de mi modelo en datos no vistos, podría considerar ajustar los hiperparámetros, aumentar el tamaño del conjunto de datos si es posible o explorar diferentes algoritmos de regresión.

## 2. Agrega al conjunto de datos columnas que representen los cuadrados de las variables predictoras (por ejemplo, M2, W2), así como los productos entre pares de variables (por ejemplo, PxS, MxW). Evalúa un modelo perceptrón multicapa para este nuevo conjunto de datos.


In [ ]:
df.head()

VR     M     W     H     P     S
0   761  41.8  75.2  86.6   9.1  14.3
1   780  67.4  73.5  66.9  17.4  11.5
2   593  44.7  82.9  66.3  20.0  10.7
3   715  84.7  88.6  78.7  15.4  12.1
4  1078  96.7  79.3  76.2  18.2  12.5

In [ ]:
import pandas as pd
import numpy as np

# Definir las variables originales
x_variables = ['M', 'W', 'H', 'P', 'S']

# Extraer las variables predictoras y la variable de respuesta
X = df[x_variables].values
y = df['VR'].values

# Agregar las variables al cuadrado
X_squared = X ** 2

# Extender la lista de nombres x_variables con los nombres de las variables al cuadrado
x_squared_variable_names = [variable + '_squared' for variable in x_variables]
x_variables.extend(x_squared_variable_names)

# Combinar X y X_squared
X = np.column_stack((X, X_squared))

# Comparar 'M' con 'M_squared'
comparison_df = pd.DataFrame({'M': X[:, x_variables.index('M')], 'M_squared': X[:, x_variables.index('M_squared')]})

print(comparison_df.head())

      M  M_squared
0  41.8    1747.24
1  67.4    4542.76
2  44.7    1998.09
3  84.7    7174.09
4  96.7    9350.89


In [ ]:
# Agregar interacciones solo entre las variables originales
new_variable_names = []

num_predictors = X.shape[1]
num_original_variables = len(x_variables)

for i in range(num_original_variables):
    for j in range(i + 1, num_original_variables):
        # Verificar que ambas variables involucradas en la interacción sean originales
        if (x_variables[i] in x_variables[:5]) and (x_variables[j] in x_variables[:5]):
            interaction_term = X[:, i] * X[:, j]
            interaction_term_name = x_variables[i] + 'x' + x_variables[j]
            X = np.column_stack((X, interaction_term))
            new_variable_names.append(interaction_term_name)

# Extender x_variables con las variables al cuadrado
x_variables.extend(new_variable_names)

In [ ]:
x_variables

['M',
 'W',
 'H',
 'P',
 'S',
 'M_squared',
 'W_squared',
 'H_squared',
 'P_squared',
 'S_squared',
 'MxW',
 'MxH',
 'MxP',
 'MxS',
 'WxH',
 'WxP',
 'WxS',
 'HxP',
 'HxS',
 'PxS']

In [ ]:
# Crear un DataFrame con las columnas adecuadas
column_names = ['M_squared', 'W_squared', 'H_squared', 'P_squared', 'S_squared',
                'MxW', 'MxH', 'MxP', 'WxH', 'WxP', 'HxP','WxS']

# Tomar solo las primeras 10 columnas de X que coinciden con las especificadas en column_names
X_df = pd.DataFrame(X[:, :12], columns=column_names)

# Concatenar las columnas de X_df a df
df = pd.concat([df, X_df], axis=1)
df.columns

Index(['VR', 'M', 'W', 'H', 'P', 'S', 'M_squared', 'W_squared', 'H_squared',
       'P_squared', 'S_squared', 'MxW', 'MxH', 'MxP', 'WxH', 'WxP', 'HxP',
       'WxS'],
      dtype='object')

In [ ]:
y = df.iloc[:, 0]
x = df.iloc[:, 1:]
print(x.head())


      M     W     H     P     S  M_squared  W_squared  H_squared  P_squared  \
0  41.8  75.2  86.6   9.1  14.3       41.8       75.2       86.6        9.1   
1  67.4  73.5  66.9  17.4  11.5       67.4       73.5       66.9       17.4   
2  44.7  82.9  66.3  20.0  10.7       44.7       82.9       66.3       20.0   
3  84.7  88.6  78.7  15.4  12.1       84.7       88.6       78.7       15.4   
4  96.7  79.3  76.2  18.2  12.5       96.7       79.3       76.2       18.2   

   S_squared      MxW      MxH      MxP     WxH     WxP      HxP      WxS  
0       14.3  1747.24  5655.04  7499.56   82.81  204.49  3143.36  3619.88  
1       11.5  4542.76  5402.25  4475.61  302.76  132.25  4953.90  4509.06  
2       10.7  1998.09  6872.41  4395.69  400.00  114.49  3705.63  2963.61  
3       12.1  7174.09  7849.96  6193.69  237.16  146.41  7504.42  6665.89  
4       12.5  9350.89  6288.49  5806.44  331.24  156.25  7668.31  7368.54  


In [ ]:

# Fit regression model using the complete dataset and calculate MSE and MAE
regr = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
regr.fit(x, y)

y_pred = regr.predict(x)
print('MSE: ', mean_squared_error(y, y_pred))
print("MAE: ", mean_absolute_error(y, y_pred))

# Evaluate regression model using k-fold cross-validation
n_folds = 3
# Set a random seed for reproducibility
np.random.seed(42)
kf = KFold(n_splits=n_folds, shuffle = True,  random_state=42)

mse_cv = []
mae_cv = []
for train_index, test_index in kf.split(x):

    # Training phase
    x_train = x.iloc[train_index]
    y_train = y.iloc[train_index]

    regr_cv = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
    regr_cv.fit(x_train, y_train)

    # Test phase
    x_test = x.iloc[test_index]
    y_test = y.iloc[test_index]

    y_pred = regr_cv.predict(x_test)

    # Calculate MSE and MAE

    mse_i = mean_squared_error(y_test, y_pred)
    print('mse = ', mse_i)
    mse_cv.append(mse_i)

    mae_i = mean_absolute_error(y_test, y_pred)
    print('mae = ', mae_i)
    mae_cv.append(mae_i)



# Calculate and print the average MSE and MAE across all folds
avg_mse = np.mean(mse_cv)
print(f"Average MSE across all folds: {avg_mse}")
avg_msa = np.mean(mae_cv)
print(f"Average MAE across all folds: {avg_msa}")


MSE:  6132.276831769321
MAE:  58.05682596660111
mse =  102414.72706665014
mae =  265.4129365782351
mse =  137069.829203595
mae =  204.2237416641957
mse =  191892.07500940587
mae =  330.4213680661957
Average MSE across all folds: 143792.21042655033
Average MAE across all folds: 266.68601543620883


comparamos con modelo lineal

In [ ]:
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.linear_model import LinearRegression
# Crear una instancia del modelo de regresión lineal
model = LinearRegression()

# Calcular las predicciones usando validación cruzada de k-pliegues
predictions = cross_val_predict(model, x, y, cv=5)  # Cambia el valor de cv según el número de pliegues deseado

# Calcular métricas de evaluación
r2 = r2_score(y, predictions)
mse = mean_squared_error(y, predictions)
mae = mean_absolute_error(y, predictions)

# Imprimir las métricas
print("Coeficiente de determinación (R2):", r2)
print("Error cuadrático medio (MSE):", mse)
print("Error absoluto medio (MAE):", mae)

Coeficiente de determinación (R2): 0.6859513135149171
Error cuadrático medio (MSE): 59906.17088630082
Error absoluto medio (MAE): 160.0346846678224


## 3. Viendo los resultados de regresión, desarrolla una conclusión sobre los siguientes puntos:

**¿Consideras que el modelo perceptrón multicapa es efectivo para modelar los datos del problema? ¿Por qué?**
Analizando los resultados y las variables no creo que estos datos son los adecuados para usar un modelo de perceptrón multicapa. Esto es porque el modelo se utiliza comúnmente en problemas de regresión en los que las relaciones entre las variables no pueden describirse de manera lineal. Analizando las variables independientes con la variable "y", si exsite una relación lineal en la mayoría de ellas. Esto se puede visualizar facilmente con unos qqplots comparando la variable "VR" con cada variable independiente.


**¿Qué modelo es mejor para los datos de criminalidad, el lineal o el perceptrón multicapa? ¿Por qué?**

Como mencione anteriormente, dado que la mayoría de las variables tienen una relación lineal con la variable "y", creo que para los datos de criminalidad si es mejor utilizar un modelo lineal. Comparando los resultado incluso el modelo lineal tiene un ligero mejor rendimineto y si lo ajustamos y lo mejoramos estoy seguro que pudieramos mejorar aun mas su rendimiento. En conclusión, para este conjunto de datos si es mejor el modelo lineal.